File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public String solution(String[] survey, int[] choices) {
        Map<Character, Integer> score = new HashMap<>();
        score.put('R', 0);
        score.put('T', 0);
        score.put('C', 0);
        score.put('F', 0);
        score.put('J', 0);
        score.put('M', 0);
        score.put('A', 0);
        score.put('N', 0);
        
        for (int i = 0; i < survey.length; i++) {
            if (choices[i] == 4) continue;
            
            int point = Math.abs(choices[i] - 4);
            if (choices[i] < 4) {
                char c = survey[i].charAt(0);
                score.put(c, score.get(c) + point);
            } else if (choices[i] > 4) {
                char c = survey[i].charAt(1);
                score.put(c, score.get(c) + point);
            }
        }
        
        StringBuilder sb = new StringBuilder();
        if (score.get('R') >= score.get('T')) {
            sb.append('R');
        } else {
            sb.append('T');
        }
        
        if (score.get('C') >= score.get('F')) {
            sb.append('C');
        } else {
            sb.append('F');
        }
        
        if (score.get('J') >= score.get('M')) {
            sb.append('J');
        } else {
            sb.append('M');
        }
        
        if (score.get('A') >= score.get('N')) {
            sb.append('A');
        } else {
            sb.append('N');
        }
        
        return sb.toString();
    }
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.kt
class Solution {
    fun solution(survey: Array<String>, choices: IntArray): String {
        val score = IntArray(26)
        val points = intArrayOf(0, 3, 2, 1, 0, 1, 2, 3)
        
        for (i in survey.indices) {
            val left = survey[i][0]
            val right = survey[i][1]
            val choice = choices[i]
            when {
                choice < 4 -> score[left.code - 'A'.code] += points[choice]
                choice > 4 -> score[right.code - 'A'.code] += points[choice]
                else -> Unit
            }
        }
        
        val pairs = arrayOf('R' to 'T', 'C' to 'F', 'J' to 'M', 'A' to 'N')
        val sb = StringBuilder()
        for ((a, b) in pairs) {
            val scoreA = score[a.code - 'A'.code]
            val scoreB = score[b.code - 'A'.code]
            sb.append(if (scoreA >= scoreB) a else b)
        }
        
        return sb.toString()
    }
}